examples/03_Batching/inference-batcher.cc

```diff
@@ -0,0 +1,398 @@
+                DLOG(INFO) << "triggering callback on held receive context";
+                search->second(true);
+                DLOG(INFO) << "callback completed";
+                return true;
+            }
+
+            bool StateCloseStreamDone(bool ok)
+            {
+                if (!ok)
+                    return Fail();
+                DLOG(INFO) << "closed client stream for writing";
+                ReadNext();
+                return true;
+            }
+
+            bool StateFinishedDone(bool ok)
+            {
+                if (m_Status.ok())
+                    DLOG(INFO) << "ClientContext: " << Tag() << " finished with OK";
+                else
+                    DLOG(INFO) << "ClientContext: " << Tag() << " finished with CANCELLED";
+                m_NextState = &Call::StateInvalid;
+                LOG(INFO) << "Batch Forwarding Completed for Tag " << Tag();
+                return false;
+            }
+
+            bool StateInvalid(bool ok)
+            {
+                LOG(FATAL) << "This should never be called";
+            }
+
+          private:
+            std::queue<Request *> m_Requests;
+            std::queue<Response *> m_Responses;
+            std::map<const Response *, Callback> m_CallbackByResponse;
+
+            bool (Call::*m_NextState)(bool);
+
+            ::grpc::Status m_Status;
+            ::grpc::ClientContext m_Context;
+            std::unique_ptr<::grpc::ClientAsyncReaderWriter<Request, Response>> m_Stream;
+            bool m_Started;
+
+            friend class Client;
+        };
+
+        void ProgressEngine(::grpc::CompletionQueue &cq)
+        {
+            void *tag;
+            bool ok = false;
+
+            while (cq.Next(&tag, &ok))
+            {
+                CHECK(ok) << "not ok";
+                Call *call = static_cast<Call *>(tag);
+                if (!call->RunNextState(ok))
+                {
+                    DLOG(INFO) << "Deleting Stream: " << tag;
+                    delete call;
+                }
+            }
+        }
+
+        int m_CurrentCQ;
+        std::unique_ptr<typename ServiceType::Stub> m_Stub;
+        std::shared_ptr<ThreadPool> m_ThreadPool;
+        std::vector<std::unique_ptr<::grpc::CompletionQueue>> m_CQs;
+    };
+
+    class Resources : public ::yais::Resources
+    {
+      public:
+        Resources(uint32_t max_batch_size, uint64_t timeout, std::shared_ptr<Client> client)
+            : m_MaxBatchsize(max_batch_size), m_Timeout(timeout), m_Client(client) {}
+
+        void Push(Request *req, Response *resp, Callback callback)
+        {
+            // thread_local ProducerToken token(m_MessageQueue);
+            // m_MessageQueue.enqueue(token, MessageType(req, resp, callback));
+            m_MessageQueue.enqueue(MessageType{req, resp, callback});
+        }
+
+        void ProgressEngine()
+        {
+            const uint64_t quanta = 100;
```

Comment:
Since these can be evaluated at compile time, would you consider `constexpr`?
```cpp
constexpr uint64_t quanta = 100;
constexpr double timeout = static_cast<double>(m_Timeout - quanta) / 1000000.0;
```

Reply:
Nice!

I learn a few more best practices everyday!  Keep em coming ;-)

Reply:
`m_Timeout` is a variable member, so `this` is not a constant expression.